jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
First things first: it's going to be a long comment again, sorry about that.

During the weekend, I've tried to implement the variant I've proposed. Actually, I did. And it worked out almost perfectly, but only almost. It turned out (nothing new but to me) that on ARM32, returning 64 bit structs is not the same as returning 64 bit ints or floats (no 64 bit struct return in register pairs, but via memory only). And this blew up code size. Moving one of the components of the previous result struct into an out-parameter (i.e., returning the number of consumed JS arguments via `return` and writing an error value to a pointer passed in as a function parameter) didn't solve things either: it has caused more than 4 function parameters in some cases, which triggered parameter passing via stack (for some params), which was not good for code size either.

So much about my failures.

But the iterator in its current form was still on my "don't like" list. I wanted to figure out why and then I realized (at least, I think so, now). Iterators (conceptually) are for iterating over the elements of a container (sorry for repeating the obvious). But the current iterator mixes things. It contains both the `c_arg_p` AND the data structure needed for iterating over the `js_arg_p` array. But they are two separate things. IMO. The `c_arg_p` array is nothing that the transform functions should be allowed to fiddle with. And this is quite obvious in `jerryx_arg_transform_args`, which is forced to fix up `iterator.c_arg_p` in every iteration of its loop because it cannot be sure that the transform functions don't mess it up.

So, I've created a "stricter" variant of the iterator, which contained the data structure necessary to iterate over the JS args only. And also changed the interface of the transform functions to take the native arg as a separate parameter.  The key change is:

```c
typedef struct
{
  const jerry_value_t *js_arg_p;
  jerry_length_t js_arg_cnt;
} jerryx_arg_js_iterator_t;

/**
 * Signature of the transform function.
 */
typedef jerry_value_t (*jerryx_arg_transform_func_t) (jerryx_arg_js_iterator_t *js_args, /**< available JS args */
                                                      const jerryx_arg_t *c_arg_p); /**< native arg */
```

The whole code is available at https://github.com/akosthekiss/jerryscript/commit/be3278698d4bf78adcf4a435abb1ddb6c69bd554 (branch name pr1740-iter). (There are still some style glitches in the commit, but it passes all CI tests.)

IMO this has resulted cleaner code. Another effect is that `c_arg_p` fields don't have to be accessed through double indirection (`iterator_p->c_arg_p->dest`). Yet another effect is that on all platforms where I got the chance to measure code size, it got smaller (15-16% on x86_64, 8% on ARM32).

Measurement details and results are below:

```bash
$ uname -mrs
Darwin 16.5.0 x86_64

$ git checkout pr1740
$ git show --oneline --quiet
ef20483c A tool in jerry-ext: arg transformer for binding
$ ./tools/build.py --clean --lto=off --jerry-ext=on
$ size ./build/lib/libjerry-ext.a
__TEXT	__DATA	__OBJC	others	dec	hex
222	0	0	96	318	13e	./build/lib/libjerry-ext.a(arg-iterator-helper.c.o)
2417	0	0	864	3281	cd1	./build/lib/libjerry-ext.a(arg-transform-functions.c.o)
422	0	0	64	486	1e6	./build/lib/libjerry-ext.a(args.c.o)

$ git checkout pr1740-iter
$ git show --oneline --quiet
be327869 Experiment: jerry-ext/args with stricter iterators
$ ./tools/build.py --clean --lto=off --jerry-ext=on
$ size ./build/lib/libjerry-ext.a
__TEXT	__DATA	__OBJC	others	dec	hex
2328	0	0	672	3000	bb8	./build/lib/libjerry-ext.a(arg-transform-functions.c.o)
392	0	0	64	456	1c8	./build/lib/libjerry-ext.a(args.c.o)

$ uname -mrs
Linux 3.13.0-63-generic x86_64

$ git checkout pr1740
$ git show --oneline --quiet
ef20483 A tool in jerry-ext: arg transformer for binding
$ ./tools/build.py --clean --lto=off --jerry-ext=on
$ size ./build/lib/libjerry-ext.a
text	   data	    bss	    dec	    hex	filename
 420	      0	      0	    420	    1a4	args.c.o (ex ./build/lib/libjerry-ext.a)
 149	      0	      0	    149	     95	arg-iterator-helper.c.o (ex ./build/lib/libjerry-ext.a)
1996	      0	      0	   1996	    7cc	arg-transform-functions.c.o (ex ./build/lib/libjerry-ext.a)

$ git checkout pr1740-iter
$ git show --oneline --quiet
be32786 Experiment: jerry-ext/args with stricter iterators
$ ./tools/build.py --clean --lto=off --jerry-ext=on
$ size ./build/lib/libjerry-ext.a
text	   data	    bss	    dec	    hex	filename
 399	      0	      0	    399	    18f	args.c.o (ex ./build/lib/libjerry-ext.a)
1751	      0	      0	   1751	    6d7	arg-transform-functions.c.o (ex ./build/lib/libjerry-ext.a)

$ arm-linux-gnueabihf-gcc --version
arm-linux-gnueabihf-gcc (Ubuntu/Linaro 4.8.4-2ubuntu1~14.04.1) 4.8.4

$ git checkout pr1740
$ git show --oneline --quiet
ef20483 A tool in jerry-ext: arg transformer for binding
$ ./tools/build.py --clean --lto=off --jerry-ext=on --toolchain=cmake/toolchain_linux_armv7l.cmake
$ arm-linux-gnueabihf-size ./build/lib/libjerry-ext.a
   text	   data	    bss	    dec	    hex	filename
    169	      0	      0	    169	     a9	args.c.o (ex ./build/lib/libjerry-ext.a)
     52	      0	      0	     52	     34	arg-iterator-helper.c.o (ex ./build/lib/libjerry-ext.a)
    938	      0	      0	    938	    3aa	arg-transform-functions.c.o (ex ./build/lib/libjerry-ext.a)

$ git checkout pr1740-iter
$ git show --oneline --quiet
be32786 Experiment: jerry-ext/args with stricter iterators
$ ./tools/build.py --clean --lto=off --jerry-ext=on --toolchain=cmake/toolchain_linux_armv7l.cmake
$ arm-linux-gnueabihf-size ./build/lib/libjerry-ext.a
text	   data	    bss	    dec	    hex	filename
 173	      0	      0	    173	     ad	args.c.o (ex ./build/lib/libjerry-ext.a)
 890	      0	      0	    890	    37a	arg-transform-functions.c.o (ex ./build/lib/libjerry-ext.a)
```